test/fw/ptl/lib/pbs_testlib.py

```diff
@@ -10929,6 +10886,48 @@ def log_match(self, msg=None, id=None, n=50, tail=True, allmatch=False,
                                max_attempts, interval, starttime, endtime,
                                level=level, existence=existence)
 
+    def run_scheduling_cycle(self):
+        """
+        Convenience method to start and finish a sched cycle
+
+        :return start time of the new sched cycle
+        """
+        sched = self.attributes['id']
+        old_val = self.server.status(SCHED, 'scheduling', id=sched)[
+            0]['scheduling']
+
+        # Make sure that we aren't in a sched cycle already
+        self.server.manager(MGR_CMD_SET, SCHED, {
+                            'scheduling': 'False'}, id=sched)
+
+        # Kick a new cycle
+        tbefore = time.time()
+        self.server.manager(MGR_CMD_SET, SCHED, {
+                            'scheduling': 'True'}, id=sched)
+        msg = self.log_match("Starting Scheduling",
+                             starttime=tbefore)
+        self.log_match("Leaving Scheduling",
+                       starttime=tbefore, interval=1,
+                       max_attempts=1200)
```

Comment:
This is not the most efficient way to check if the scheduling cycle is over.  Do an expect() on scheduling NE 'scheduling'.  That will limit the amount of string comparisons.

Reply:
I was going for reliability over performance. I didn't do what you are suggesting because PBS might execute multiple sched cycles, so i thought that checking for the leaving sched cycle log message with starttime set correctly would be more reliable than checking for the scheduling attribute. I wish we could enhance PBS to run 'n' sched cycles, that way we could just run 1 cycle reliably, but that doesn't really make much sense for end users.

Reply:
While you are correct, your method will definitely catch the one cycle, I'm not sure it matters.  In situation 1, we check if scheduler's state != scheduling.  We miss our current cycle, and catch a subsequent one.  We eventually return.  In situation 2, run_cycle() makes sure that it matches the end of the first cycle.  It then returns back to the caller.  By the time we get control back, the scheduler has run another one.  In both cases we have run two cycles.  The reason I like checking for state != scheduling is that manager() does that.  If we went with log_match() we'd have two functions checking for the same thing in two different ways.